src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,190 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] a = list.toArray();
+        int numNew = a.length;
+        if (numNew == 0) {
+            return;
+        }
+        Node<T> pred;
+        pred = tail;
+        for (Object o : a) {
+            T e = (T) o;
+            Node<T> newNode = new Node<>(pred, e, null);
+            if (pred == null) {
+                head = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            pred = newNode;
+        }
+        tail = pred;
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0) {
+            checkIndex(index);
+        } else {
+            checkIndex(index + 1);
+        }
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0) {
+            checkIndex(index);
+        } else {
+            checkIndex(index + 1);
+        }
+        Node<T> x = node(index);
+        T oldVal = x.item;
+        x.item = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0) {
+            checkIndex(index);
+        } else {
+            checkIndex(index + 1);
+        }
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
```

Comment:
this method can be shortened, use single loop for iteration. Inside use the same check you used in arr list and storage